### PR TITLE
chore: Run pyupgrade on codebase (3.6+)

### PR DESCRIPTION
**Pull request by kratsg**
# Description

Run [asottile/pyupgrade](https://github.com/asottile/pyupgrade) on our codebase. Not sure how to implement a CI/github actions for this (@asottile might have suggestions?)

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Run pyupgrade --py36-plus on codebase
* Add pyupgrade to pre-commit hooks
* Add pre-commit.ci status badge to README
* Use f-strings over format() as much as possible
* Thanks to Anthony Sottile (@asottile) for pyupgrade and pre-commit.ci

Co-authored-by: Matthew Feickert <matthew.feickert@cern.ch>
```

### CODE REVIEW

**Review by lukasheinrich — APPROVED** [review]

**Comment by asottile** [comment]
for CI suggestions it looks like you're already using [pre-commit](https://github.com/scikit-hep/pyhf/blob/master/.pre-commit-config.yaml) so you can [add this configuration there](https://github.com/asottile/pyupgrade#as-a-pre-commit-hook)

as for running in CI, I've been building https://pre-commit.ci -- if you're interested in the open-but-secret alpha you can set it up at https://results.pre-commit.ci

otherwise if you want to stick to github actions there's https://github.com/pre-commit/action

File: src/pyhf/modifiers/__init__.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def add_to_registry(
         tensorlib, _ = get_backend()
         if not hasattr(tensorlib, pdf_type):
             raise exceptions.InvalidModifier(
-                'The specified pdf_type "{0:s}" is not valid for {1:s}({2:s}). See pyhf.tensor documentation for available pdfs.'.format(
+                'The specified pdf_type "{:s}" is not valid for {:s}({:s}). See pyhf.tensor documentation for available pdfs.'.format(
```
Comment (matthewfeickert) [inline src/pyhf/modifiers/__init__.py:48]: @kratsg Can we go in and get these should be f-strings by hand?

File: src/pyhf/modifiers/__init__.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def add_to_registry(
 
     if op_code not in ['addition', 'multiplication']:
         raise exceptions.InvalidModifier(
-            'The specified op_code "{0:s}" is not valid for {1:s}({2:s}). See pyhf.modifier documentation for available operation codes.'.format(
+            'The specified op_code "{:s}" is not valid for {:s}({:s}). See pyhf.modifier documentation for available operation codes.'.format(
```
Comment (matthewfeickert) [inline src/pyhf/modifiers/__init__.py:58]: c.f. above.

File: src/pyhf/modifiers/__init__.py [hunk 3]
```diff
@@ -168,7 +168,7 @@ def wrapper(cls):
         return args[0]
     else:
         raise ValueError(
-            '@modifier must be called with only keyword arguments, @modifier(name=\'foo\'), or no arguments, @modifier; ({0:d} given)'.format(
+            '@modifier must be called with only keyword arguments, @modifier(name=\'foo\'), or no arguments, @modifier; ({:d} given)'.format(
```
Comment (matthewfeickert) [inline src/pyhf/modifiers/__init__.py:171]: c.f. above.

File: src/pyhf/pdf.py [hunk 4]
```diff
@@ -26,7 +26,7 @@ def _paramset_requirements_from_channelspec(spec, channel_nbins):
         for sample in channel['samples']:
             if len(sample['data']) != channel_nbins[channel['name']]:
                 raise exceptions.InvalidModel(
-                    'The sample {0:s} has {1:d} bins, but the channel it belongs to ({2:s}) has {3:d} bins.'.format(
+                    'The sample {:s} has {:d} bins, but the channel it belongs to ({:s}) has {:d} bins.'.format(
```
Comment (matthewfeickert) [inline src/pyhf/pdf.py:29]: Manually get f-strings?

File: src/pyhf/pdf.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def _paramset_requirements_from_channelspec(spec, channel_nbins):
                     ].required_parset(sample['data'], modifier_def['data'])
                 except KeyError:
                     log.exception(
-                        'Modifier not implemented yet (processing {0:s}). Available modifiers: {1}'.format(
+                        'Modifier not implemented yet (processing {:s}). Available modifiers: {}'.format(
```
Comment (matthewfeickert) [inline src/pyhf/pdf.py:45]: Manually get f-strings?

File: src/pyhf/pdf.py [hunk 6]
```diff
@@ -298,7 +298,7 @@ def suggested_fixed(self):
     def set_poi(self, name):
         if name not in [x for x, _ in self.modifiers]:
             raise exceptions.InvalidModel(
-                "The parameter of interest '{0:s}' cannot be fit as it is not declared in the model specification.".format(
+                "The parameter of interest '{:s}' cannot be fit as it is not declared in the model specification.".format(
```
Comment (matthewfeickert) [inline src/pyhf/pdf.py:301]: Manually get f-strings?

File: src/pyhf/readxml.py [hunk 7]
```diff
@@ -80,7 +80,7 @@ def process_sample(
 
     for modtag in modtags:
         modtags.set_description(
-            '  - modifier {0:s}({1:s})'.format(
+            '  - modifier {:s}({:s})'.format(
```
Comment (matthewfeickert) [inline src/pyhf/readxml.py:83]: Manually get f-strings?

File: src/pyhf/readxml.py [hunk 8]
```diff
@@ -312,7 +312,7 @@ def dedupe_parameters(parameters):
             for p in parameter_list:
                 log.warning(p)
             raise RuntimeError(
-                'cannot import workspace due to incompatible parameter configurations for {0:s}.'.format(
+                'cannot import workspace due to incompatible parameter configurations for {:s}.'.format(
```
Comment (matthewfeickert) [inline src/pyhf/readxml.py:315]: Manually get f-strings?

File: src/pyhf/workspace.py [hunk 9]
```diff
@@ -349,12 +348,12 @@ def get_measurement(
             if measurement_name is not None:
                 if measurement_name not in self.measurement_names:
                     log.debug(
-                        'measurements defined:\n\t{0:s}'.format(
+                        'measurements defined:\n\t{:s}'.format(
```
Comment (matthewfeickert) [inline src/pyhf/workspace.py:351]: Manually get f-strings?

File: src/pyhf/workspace.py [hunk 10]
```diff
@@ -349,12 +348,12 @@ def get_measurement(
...
                             '\n\t'.join(self.measurement_names)
                         )
                     )
                     raise exceptions.InvalidMeasurement(
-                        'no measurement by name \'{0:s}\' was found in the workspace, pick from one of the valid ones above'.format(
+                        'no measurement by name \'{:s}\' was found in the workspace, pick from one of the valid ones above'.format(
```
Comment (matthewfeickert) [inline src/pyhf/workspace.py:356]: Manually get f-strings?

File: src/pyhf/writexml.py [hunk 11]
```diff
@@ -209,7 +209,7 @@ def build_modifier(spec, modifierspec, channelname, samplename, sampledata):
         pass
     else:
         log.warning(
-            'Skipping modifier {0}({1}) for now'.format(
+            'Skipping modifier {}({}) for now'.format(
```
Comment (matthewfeickert) [inline src/pyhf/writexml.py:212]: Manually get f-strings?

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
Nice. @kratsg Can we also get all these "should be f-string" cases taken care of as well?

**Comment by matthewfeickert** [comment]
> as for running in CI, I've been building https://pre-commit.ci -- if you're interested in the open-but-secret alpha you can set it up at https://results.pre-commit.ci

Thanks @asottile! I've gone ahead and setup `pyhf` on https://pre-commit.ci :+1:

**Comment by asottile** [comment]
uh hmmm that's interesting -- it's not supposed to double-update so I've got a bug to fix (the recursion detection appears broken!)

**Review by matthewfeickert — APPROVED** [review]
@kratsg Actually, if we want to just get the f-string stuff in a seperate PR I'm good with that so that we can just move on with this nice change. :) Thanks for making great stuff @asottile.

**Pull request merged**